dig.go

```diff
@@ -565,11 +593,61 @@ func (cv connectionVisitor) Visit(res result) resultVisitor {
 		// value there.
 		k := key{group: r.Group, t: r.Type}
 		cv.keyPaths[k] = path
+
+	case resultAs:
+		k := key{name: r.Name, t: r.Type}
+		if err := checkPathsForKey(cv.keyPaths, k, path); err != nil {
+			*cv.err = err
+			return nil
+		}
+		if err := checkProvidersForKey(cv.c.providers, k, path); err != nil {
+			*cv.err = err
+			return nil
+		}
+		cv.keyPaths[k] = path
+
+		for _, as := range r.As {
+			asIndirectValue := reflect.Indirect(reflect.ValueOf(as))
+			k := key{name: r.Name, t: asIndirectValue.Type()}
+
+			if err := checkPathsForKey(cv.keyPaths, k, path); err != nil {
+				*cv.err = err
+				return nil
+			}
+			if err := checkProvidersForKey(cv.c.providers, k, path); err != nil {
+				*cv.err = err
+				return nil
+			}
+			cv.keyPaths[k] = path
+		}
```

Comment:
I think we can dry this up a little. These helpers closely resemble the helpers used in the `resultSingle` case. We can consolidate the helper functions into a method on the connectionVisitor:
```
func (cv connectionVisitor) checkKey(k key, path string) error {
	if err := checkPathsForKey(cv.keyPaths, k, path); err != nil {
		return err
	}
	if err := checkProvidersForKey(cv.c.providers, k, path); err != nil {
		return err
	}
	return nil
}
```
Since we are only calling these helpers from this function, we can flatten their implementations here, too. This would reduce the layers of indirection slightly.
```
func (cv connectionVisitor) checkKey(k key, path string) error {
	if conflict, ok := cv.keyPaths[k]; ok {
		return fmt.Errorf(
			"cannot provide %v from %v: already provided by %v",
			k, path, conflict)
	}
	if ps := cv.c.providers[k]; len(ps) > 0 {
		cons := make([]string, len(ps))
		for i, p := range ps {
			cons[i] = fmt.Sprint(p.Location())
		}

 		return fmt.Errorf(
			"cannot provide %v from %v: already provided by %v",
			k, path, strings.Join(cons, "; "))
	}
	return nil
}
```
This way, we can call this as:
```
	case resultAs:
		k := key{name: r.Name, t: r.Type}
		if err := cv.checkKey(k, path); err != nil {
			*cv.err = err
			return nil
		}
		cv.keyPaths[k] = path

 		for _, as := range r.As {
			asIndirectValue := reflect.Indirect(reflect.ValueOf(as))
			k := key{name: r.Name, t: asIndirectValue.Type()}

 			if err := cv.checKey(k, path); err != nil {
				*cv.err = err
				return nil
			}
			cv.keyPaths[k] = path
		}
	}

	return cv
```

Now we could use this function to `resultSingle`, too.

Reply:
Sound good to me :-)

When I was working on it I thought to group those those two calls in a single function but wasn't sure how to call it - `checkKey` sounds a good name.